### ISSUE TITLE
Design an initial IPC mechanism for Arrow Vectors

### ISSUE THREAD

**Issue by toddfarmer** [toddfarmer/arrow-migration#1684]
***Note**: This issue was originally created as [ARROW-263](https://issues.apache.org/jira/browse/ARROW-263). Please see the [migration documentation](https://gist.github.com/toddfarmer/12aa88361532d21902818a6044fda4c3) for further details.*

### Original Issue Description:
Prior discussion on this topic [1].

Use-cases:
1.  User defined function (UDF) execution:  One process wants to execute a user defined function written in another language (e.g. Java executing a function defined in python, this involves creating Arrow Arrays in java, sending them to python and receiving a new set of Arrow Arrays produced in python back in the java process).
2.  If a storage system and a query engine are running on the same host we might want use IPC instead of RPC (e.g. Apache Drill querying Apache Kudu)

Assumptions:
1.  IPC mechanism should be useable from the core set of supported languages (Java, Python, C) on POSIX and ideally windows systems.  Ideally, we would not need to add dependencies on additional libraries outside of each languages outside of this document.
We want leverage shared memory for Arrays to avoid doubling RAM requirements by duplicating the same Array in different memory locations.  
2. Under some circumstances shared memory might be more efficient than FIFOs or sockets (in other scenarios they won’t see thread below).
3. Security is not a concern for V1, we assume all processes running are “trusted”.

Requirements:
1.Resource management: 
    a.  Both processes need a way of allocating memory for Arrow Arrays so that data can be passed from one process to another.
    b. There must be a mechanism to cleanup unused Arrow Arrays to limit resource usage but avoid race conditions when processing arrays
2.  Schema negotiation - before sending data, both processes need to agree on schema each one will produce.

Out of scope requirements:
1.  IPC channel metadata discovery is out of scope of this document.  Discovery can be provided by passing appropriate command line arguments, configuration files or other mechanisms like RPC (in which case RPC channel discovery is still an issue).

[1] http://mail-archives.apache.org/mod_mbox/arrow-dev/201603.mbox/%3C8D5F7E3237B3ED47B84CF187BB17B666148E7322@SHSMSX103.ccr.corp.intel.com%3E

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-263?focusedCommentId=15427516) by Micah Kornfield (emkornfield@gmail.com):*
TL;DR:  Current rough proposal/thoughts:
1.  used memory mapped files (clients of the library should use a directory backed by an in memory file system, e.g. TMPFS, but for debugging it could be useful to use traditional file system).  
2.  An initial version of IPC will focus on one way, producer->consumer channel.  UDF execution can be handled by creating two channels (one in each direction from the process).
3.  It is still an open question, but trying to do IPC purely though shared memory seems like a more difficult approach compared to a traditional RPC (e.g. thrift/GRPC) but more complex.


More in depth analysis based on my research:

On POSIX systems there are a two core APIs to consider from the C++ implementation side:
1.  Shared Memory APIs [1]:  These create shared memory objects.  Shared memory objects are named and persist after a process terminates (but not after a system restart).  The APIs for manipulating shared memory return a file descriptor that is MMAPPED.
2.  MMAP APIs [2]: These take either a file descriptor (either shared memory descriptor or a traditional file descriptor) and map the contents of the object into the processes memory space (an option exists that is anonymous MMAP, but this is only useful for sharing memory between forked processes, that don't run execve, which doesn't conform to our use-case).

 "Shared memory" vs traditional file system:
1.  From what I can tell (experiments needed), memory mapping a file created on tmpfs should have identical performance to mapping shared memory object.
2.  Both approaches suffer from the fact that if all processes crash for some reason there will be garbage left over that consumes resources\* (if a file is stored on a memory backed file system then it will get collected on a system reset).
3.  Java doesn't support shared memory natively.  It does support memory mapping files.  The behavior for when other processes see changes to files is undefined, but based on various articles it seems that at least on linux, setting up the file as read/write should be suitable for RPC (without having to call sync repeatedly)

\*In C/C++ is theoretically possible open a file (get a file descriptor) then unlink the file, then send the file descriptor over a Unix domain socket to another process.  Once all processes using the file descriptor exist all resources should get cleaned up.  Java doesn't support unix domain sockets in its core library and this solution is less likely to be as easily portable to non-posix operating systems.

[1] http://man7.org/linux/man-pages/man3/shm_open.3.html
[2] http://man7.org/linux/man-pages/man2/mmap.2.html

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-263?focusedCommentId=15427632) by Philipp Moritz (pcmoritz):*
Hey Micah,

thanks for your insights, seems we have been thinking along very similar lines. I wrote one shared memory object store based on boost::IPC, which under the hood uses shm_open, one based on Chrome's Mojo IPC system (which is great but pulls in a lot of dependencies) and currently I am writing a new very lightweight one based on POSIX. I'm happy to share my findings:

Answering your in-depth analysis questions:
1. This is essentially the design of boost::IPC and this design makes it hard to ensure that named shared objects are removed properly (they cannot be removed automatically by the OS). If the process that is supposed to clean them up crashes, you will need to delete them manually.

2. Using mmaped based APIs is the way to go in my opionion. To create the file, you create a temporary file of size 0, keep the file descriptor around, unlink the file from the file system, and manipulate the file by mmaping it from the file descriptor. If you want to share the file with another process, you pass the file descriptor over a Unix Domain Socket. On Windows there also seem to be ways of doing this.

3. I also found the guarantees given by the JVM too restrictive. One way to do it which is working for me is to use JNI and to wrap some native C code. The alternative is to use JVM memory mapped files anyways and to hope they do the right thing (which they seem to do on linux), and keep the file underlying the memory around (with similar problems as 1).

You find my implementation here: https://github.com/pcmoritz/plasma, it essentially implements what you describe in \*. I'd love to get your feedback, and if we can share some code that would be even better. I also think that Mojo (Google Chrome's new IPC layer) got this "open file of size 0, unlinke the file, keep the file descriptor around" approach working on Windows, so this should be ok.

A bit more about plasma: It allows one process to create a shared memory object, modify it, and then at some point seal it, which makes it immutable. Other processes can "get" the objects, which will block the getting thread until the object is sealed by the creator process, and then return address and size of the object.

The central question is then if people are ok with using JNI for java to implement passing filedescriptors around and mmapping the files. I have some code to do this which I'm happy to share with you in the next couple of days.

All the best,
Philipp.

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-263?focusedCommentId=15428637) by Micah Kornfield (emkornfield@gmail.com):*
`[~pcmoritz]` this looks cool.  I think JNI might be a requirement here to guarantee correctness, but I will let people doing more of the JVM development chime in.  Hopefully, somebody who develops on windows can add the equivalent work.  

Unix domain sockets also seem to suffer from the same resource cleanup issue if we want to be portable to non-linux POSIX systems.  The man-page [1] indicates normal unix domain sockets need a file path as well or you can use abstract sockets, but those are linux specific.  I see in your code you unlink the path as part of starting the server, so it seems maybe its ok for the path not to exist for the client programs to connect correctly?

One other thought.  It might nice to be able to re-use FD after the data is no longer used by either process.  Assuming roughly even size batches, I think think could avoid the page faults that would occur by creating a new shared memory/object file each time.

[1] http://man7.org/linux/man-pages/man7/unix.7.html

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-263?focusedCommentId=15429948) by Philipp Moritz (pcmoritz):*
Hey Micah,

thanks for your answer!

I got the trick of unlinking the domain socket from here: https://troydhanson.github.io/network/Unix_domain_sockets.html ("Unlink before bind"). On Linux and Mac OS it seems to work and prevents leaking of the file. Note that at some point we need to introduce a named object that can be seen by all processes to bootstrap the communication between processes and this has been the least problematic way of doing that I have seen.

At the moment I'm also working on a distributed version of the object store (with a separate process that can be used to ship objects between object stores on different nodes in a network) and investigating libuv to do it in a platform independent way. Libuv is a small dependency and my experience so far is pretty enjoyable. It also includes limited functionality to exchange file descriptors, but this might not work on windows (see also https://groups.google.com/forum/#!msg/libuv/0xxXBIGlzLc/H1HbL-igb84J, I haven't tried it yet).

Concerning your last comment: The plasma store is a long running process that keeps its file descriptor and the data alive. Are page faults still a problem if data does not need to be reloaded from hard disk?

If somebody else has a platform independent way of achieving some of these goals, I'd be happy to learn about their ideas.

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-263?focusedCommentId=15430029) by Micah Kornfield (emkornfield@gmail.com):*
Regarding the unix domain sockets.  Yes I think you are right that there needs to be a centralized named file to so that communication can be bootstrapped. 

There are two types of page faults.  Minor and Major.  Minor ones are when the pages are in memory but not mapped properly the address space of the new process yet and major ones are when the page isn't in memory at all and needs to be loaded mapped [1].  If I understood Todd Lipcon on the thread mentioned above when minor page faults are accounted for, the latency of shared memory might not be better then just transferring over unix domain sockets.

I think I have a workable design for IPC now in my head, but not sure that I will get it documented tonight, and will likely  be Away from keyboard for at least a couple of days.  

[1] https://en.wikipedia.org/wiki/Page_fault#Minor

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-263?focusedCommentId=16854525) by Antoine Pitrou (apitrou):*
Should this be kept open? It looks essentially like a brain dump and no discussion took place for the last 3 years.

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-263?focusedCommentId=16854722) by Micah Kornfield (emkornfield@gmail.com):*
I thin it can be closed.

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-263?focusedCommentId=17585822) by @toddfarmer:*
Transitioning issue from Resolved to Closed to based on resolution field value.
